In [15]:
import numpy as np
import pandas as pd
import scipy
import sklearn
import matplotlib.pyplot as plt
import seaborn as sns
from datetime import datetime

from sklearn.decomposition import PCA
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.model_selection import train_test_split
from sklearn.decomposition import TruncatedSVD
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import Normalizer
from sklearn.naive_bayes import BernoulliNB
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.linear_model import LogisticRegression
from sklearn import ensemble
from sklearn.utils import resample
from string import punctuation
import re
from sklearn.model_selection import cross_val_score

%matplotlib inline

In [16]:
reviews_raw = pd.read_json('reviews_Sports_And_Outdoors_5.JSON', lines=True)
pd.DataFrame.from_dict(reviews_raw, orient='columns')
reviews_raw.head(10)

asin helpful  overall  \
0  1881509818  [0, 0]        5   
1  1881509818  [1, 1]        5   
2  1881509818  [2, 2]        4   
3  1881509818  [0, 0]        4   
4  1881509818  [0, 0]        4   
5  1881509818  [0, 0]        5   
6  1881509818  [0, 0]        5   
7  2094869245  [0, 0]        4   
8  2094869245  [0, 1]        5   
9  2094869245  [0, 0]        5   

                                          reviewText   reviewTime  \
0  This came in on time and I am veru happy with ...  01 26, 2014   
1  I had a factory Glock tool that I was using fo...   02 2, 2012   
2  If you don't have a 3/32 punch or would like t...  02 28, 2012   
3  This works no better than any 3/32 punch you w...   02 5, 2012   
4  I purchased this thinking maybe I need a speci...  04 23, 2013   
5  Needed this tool to really break down my G22, ...   11 2, 2012   
6  If u don't have it .. Get it. All you need to ...  06 10, 2014   
7  This light will no doubt capture the attention...  08 31, 2013   
8  Light and laser torch work well, very bright. ...  05 27, 2013   
9  Does everything it says it will do. I would li...   11 2, 2013   

       reviewerID                    reviewerName  \
0    AIXZKN4ACSKI                    David Briner   
1  A1L5P841VIO02V                 Jason A. Kramer   
2   AB2W04NI4OEAD                      J. Fernald   
3  A148SVSWKTJKU6  Jusitn A. Watts "Maverick9614"   
4   AAAWJ6LW9WMOO                    Material Man   
5  A2XX2A4OJCDNLZ             RatherLiveInKeyWest   
6  A283UOBQRUNM4Q                   Thomas Dragon   
7   AWG3H90WVZ0Z1                     Alec Nelson   
8  A3V52OTJHKIJZX   A. Saenz Jr. "Bettering self"   
9  A3SZBE5F3UQ9EC               ChasRat "ChasRat"   

                                             summary  unixReviewTime  
0                                     Woks very good      1390694400  
1                  Works as well as the factory tool      1328140800  
2                          It's a punch, that's all.      1330387200  
3                    It's a punch with a Glock logo.      1328400000  
4            Ok,tool does what a regular punch does.      1366675200  
5  Glock punch tool - needed for your Glock and o...      1351814400  
6                                         Great tool      1402358400  
7                                            Bright!      1377907200  
8                                            Be seen      1369612800  
9                            Bicycle rear tail light      1383350400

In [17]:
null_count = reviews_raw.isnull().sum()
null_count[null_count>0]

reviewerName    1402
dtype: int64

In [18]:
df2 = reviews_raw.drop(['reviewerID', 'reviewerName', 'asin', 'helpful', 'reviewTime', 'summary', 
                       'unixReviewTime'], 1)
threshold = 4
df2['Sentiment'] = np.where(df2['overall'] >= threshold, 1,0)
df2 = df2.drop(['overall'], 1)

df2.head(10)

reviewText  Sentiment
0  This came in on time and I am veru happy with ...          1
1  I had a factory Glock tool that I was using fo...          1
2  If you don't have a 3/32 punch or would like t...          1
3  This works no better than any 3/32 punch you w...          1
4  I purchased this thinking maybe I need a speci...          1
5  Needed this tool to really break down my G22, ...          1
6  If u don't have it .. Get it. All you need to ...          1
7  This light will no doubt capture the attention...          1
8  Light and laser torch work well, very bright. ...          1
9  Does everything it says it will do. I would li...          1

In [19]:
df2['Sentiment'].value_counts()

1    253017
0     43320
Name: Sentiment, dtype: int64

In [20]:
# Utility function for standard text cleaning.
def text_cleaner(text):
    # Visual inspection identifies a form of punctuation spaCy does not
    # recognize: the double dash '--'.  Better get rid of it now!
    text = re.sub(r'--',' ',text)
    text = re.sub("[\[].*?[\]]", "", text)
    text = ' '.join(text.split())
    return text

for review in df2['reviewText']:
    review = text_cleaner(review)

df2.head(10)

reviewText  Sentiment
0  This came in on time and I am veru happy with ...          1
1  I had a factory Glock tool that I was using fo...          1
2  If you don't have a 3/32 punch or would like t...          1
3  This works no better than any 3/32 punch you w...          1
4  I purchased this thinking maybe I need a speci...          1
5  Needed this tool to really break down my G22, ...          1
6  If u don't have it .. Get it. All you need to ...          1
7  This light will no doubt capture the attention...          1
8  Light and laser torch work well, very bright. ...          1
9  Does everything it says it will do. I would li...          1

In [21]:
positive = df2.loc[df2['Sentiment'] == 1]
negative = df2.loc[df2['Sentiment'] == 0]
df3 = resample(positive, n_samples=43320)
negative = resample(negative, n_samples=43320)
df3 = df3.append(negative)
df3['Sentiment'].value_counts()

1    43320
0    43320
Name: Sentiment, dtype: int64

In [22]:
X = df3['reviewText']
y = df3['Sentiment']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=42)
vect2 = TfidfVectorizer(min_df=20, analyzer = 'word', stop_words = 'english', ngram_range = (1,2) ).fit(X_train)

X_train_vectorized = vect2.transform(X_train)
X_test_vectorized = vect2.transform(X_test)


In [23]:
def testing_func(model):
    
    model.fit(X_train_vectorized, y_train)
    model.fit(X_test_vectorized, y_test)

    # Inspect the results.
    print('\nR-squared for the Training Set:')
    print(model.score(X_train_vectorized, y_train))
    print('\nR-squared for the Test Set:')
    print(model.score(X_test_vectorized, y_test))
    print('Cross Validation Scores with 10 folds: {}'.format(cross_val_score(model, X_train_vectorized, y_train, cv=10)))

In [24]:
start = datetime.now()
#Let's put our variables through Naive Bayes.
bnb = BernoulliNB(alpha=1.0, binarize=0.0, class_prior=None, fit_prior=True)

testing_func(bnb)
print('It took the following time to complete this task:', datetime.now() - start)


R-squared for the Training Set:
0.728569459757

R-squared for the Test Set:
0.767172635702
Cross Validation Scores with 10 folds: [ 0.75090439  0.73953488  0.74521964  0.74091301  0.73712317  0.73540052
  0.74074074  0.74039621  0.7419466   0.74857832]
It took the following time to complete this task: 0:00:04.006573


In [25]:
start = datetime.now()
#Let's pop those into the model
lr = LogisticRegression(C=1, penalty='l1')
testing_func(lr)

print('It took the following time to complete this task:', datetime.now() - start)


R-squared for the Training Set:
0.790345920617

R-squared for the Test Set:
0.824426412983
Cross Validation Scores with 10 folds: [ 0.81119724  0.8039621   0.81498708  0.80275624  0.80086133  0.80258398
  0.79913867  0.81050818  0.8044789   0.80510081]
It took the following time to complete this task: 0:00:26.336501


In [26]:
start = datetime.now()
#Let's pop those into the model
lr = LogisticRegression(C=1, penalty='l2')
testing_func(lr)

print('It took the following time to complete this task:', datetime.now() - start)


R-squared for the Training Set:
0.801853638368

R-squared for the Test Set:
0.865871572468
Cross Validation Scores with 10 folds: [ 0.81860465  0.81119724  0.82153316  0.81619294  0.80878553  0.80620155
  0.80861326  0.81343669  0.80981912  0.81095985]
It took the following time to complete this task: 0:00:22.409759


In [27]:
start = datetime.now()
clf = DecisionTreeClassifier(max_depth=8, random_state=0)
testing_func(clf)

print('It took the following time to complete this task:', datetime.now() - start)


R-squared for the Training Set:
0.646826764057

R-squared for the Test Set:
0.663647174035
Cross Validation Scores with 10 folds: [ 0.65047373  0.64875108  0.65254091  0.64668389  0.65219638  0.64651163
  0.64254953  0.65116279  0.65064599  0.6457005 ]
It took the following time to complete this task: 0:01:40.545964


In [28]:
start = datetime.now()
#Now let's model with Gradient Boosting
clf = GradientBoostingClassifier(loss='deviance', learning_rate=1.0, n_estimators=10, subsample=1.0, 
                                          criterion='friedman_mse', min_samples_split=4, min_samples_leaf=1, 
                                          min_weight_fraction_leaf=0.0, max_depth=2, min_impurity_decrease=0.0, 
                                          min_impurity_split=None, init=None, random_state=None, max_features=None, 
                                          verbose=0, max_leaf_nodes=None, warm_start=False, presort='auto')

testing_func(clf)
print('It took the following time to complete this task:', datetime.now() - start)


R-squared for the Training Set:
0.660367282249

R-squared for the Test Set:
0.667564353665
Cross Validation Scores with 10 folds: [ 0.66890612  0.66821705  0.66425495  0.6664944   0.66614987  0.66270457
  0.6535745   0.6704565   0.66614987  0.66569016]
It took the following time to complete this task: 0:03:26.478319


Althought Bernoulli Bayes works fairly well, Ridge and Lasso Regression performed the best of all the models with around 0.80 Accuracy with similar scores on the Cross Validation/Test meaning that it likely isn't overfitting the data.